File: api/token.go
package api

import "time"

type CreateTokenRequest struct {
	Identifier     string `json:"identifier" validate:"required,min=1,max=128"`
	Type           string `json:"type"`
	OrganizationID string `json:"omitzero"`
	// Scopes optionally restricts the token to a set of "module:action"
	// capabilities (e.g. ["storage:write","logs:write"]). Empty means full access.
	Scopes []string `json:"scopes" validate:"omitempty,max=16,dive,min=1,max=64"`
	// ExpiresAt optionally sets when the token stops working. Nil means it never
	// expires.
	ExpiresAt *time.Time `json:"expiresAt"`
}

type CreateTokenResponse struct {
	Token string `json:"token"`
}

type ListTokensResponse struct {
	ID         int64      `json:"id"`
	Identifier string     `json:"identifier"`
	Scopes     []string   `json:"scopes"`
	ExpiresAt  *time.Time `json:"expiresAt"`
}

File: web/src/typings/token.ts
import { z } from "zod";

// Scopes a token can be granted. An empty selection means full access, matching
// the backend (a token with no scopes is unrestricted, for backward compat).
export const TOKEN_SCOPES = [
  { value: "storage:write", label: "Storage — upload/delete files" },
  { value: "storage:read", label: "Storage — list/read files" },
  { value: "logs:write", label: "Logs — ingest logs" },
] as const;

export type TokenScope = (typeof TOKEN_SCOPES)[number]["value"];

export const tokenSchema = z.object({
  identifier: z.string().min(1, "Identifier is required"),
  type: z.string(),
  scopes: z.array(z.string()),
  // ISO-8601 instant, or undefined for "never expires".
  expiresAt: z.string().optional(),
});

export type TokenParams = z.infer<typeof tokenSchema>;

export interface TokenPreview {
  token: string;
}

export interface Token {
  id: number;
  type: string;
  identifier: string;
  scopes?: string[] | null;
  expiresAt?: string | null;
}
